XenOvfTransport/Export.cs

```diff
@@ -48,112 +48,58 @@ public class Export
+                            if (vmgm.os_version.ContainsKey("major") && vmgm.os_version.ContainsKey("minor"))
                             {
-                                ushort osid = ValueMaps.OperatingSystem(vmgm.os_version[key]);
-                                if (osid == 0xFFFF) { osid = 1; } // change to OTHER since search failed.
-                                string version = OVF.GetContentMessage("SECTION_OPERATINGSYSTEM_INFO");
-                                if (vmgm.os_version.ContainsKey("major") &&
-                                    vmgm.os_version.ContainsKey("minor"))
-                                {
-                                    version = string.Format(OVF.GetContentMessage("SECTION_OPERATINGSYSTEM_VERSION"), vmgm.os_version["major"], vmgm.os_version["minor"]);
-                                }
-                                string osname = (vmgm.os_version[key].Split(new [] { '|' }))[0];
-                                OVF.UpdateOperatingSystemSection(ovfEnv, vsId, osname, version, osid);
-                                break;
+                                version = string.Format(OVF.GetContentMessage("SECTION_OPERATINGSYSTEM_VERSION"), vmgm.os_version["major"], vmgm.os_version["minor"]);
                             }
+
+                            string osname = vmgm.os_version[key].Split('|')[0];
+                            OVF.UpdateOperatingSystemSection(ovfEnv, vsId, osname, version, osid);
+                            break;
                         }
                     }
                 }
                 #endregion
 
-                #region ADD VSSD
-                // IS PV'd? for VirtualSystemType identification.
-                string typeformat = @"{0}-{1}-{2}";
-                string vmtype = string.Format(typeformat, "hvm", "3.0", "unknown");
-                if (vm.HVM_boot_policy != null && vm.HVM_boot_policy == Properties.Settings.Default.xenBootOptions)
-                {
-                    if (!string.IsNullOrEmpty(vm.domarch))
-                    {
-                        vmtype = string.Format(typeformat, vm.domarch, "3.0", "unknown");
-                    }
-                }
-                else
-                {
-                    if (!string.IsNullOrEmpty(vm.domarch))
-                    {
-                        vmtype = string.Format(typeformat, "xen", "3.0", vm.domarch);
-                    }
-                    else
-                    {
-                        vmtype = string.Format(typeformat, "xen", "3.0", "unknown");
-                    }
-                }
-                OVF.AddVirtualSystemSettingData(ovfEnv, vsId, vhsId, vm.name_label, OVF.GetContentMessage("VSSD_CAPTION"), vm.name_description, Guid.NewGuid().ToString(), vmtype);
+                #region ADD VirtualSystemType identification
+
+                var pv = vm.IsHVM() ? "hvm" : "xen";

```

Comment:
This is an interesting one. I think the old logic for HVM was wrong. First of all, for `string.IsNullOrEmpty(vm.domarch)` it gave `hvm-3.0-unknown` and for `!string.IsNullOrEmpty(vm.domarch)` it gave `{vm.domarch}-3.0-unknown`, which looks a bit weird, especially when compared with the PV case where we have `xen-3.0-unknown` and `xen-3.0-{vm.domarch}` respectively. Then, I'm not sure why we need to check additionally that `HVM_boot_policy` is `BIOS order` (these settings are not configurable) in order to decide whether it's HVM; it made snse to me to use the same check as elsewhere in XenCenter.

Reply:
I think your change is correct.
According to https://github.com/xenserver/xenadmin/blob/a2596ec201c6760f703833fdff1f1ea8a55548f3/XenOvfApi/app.config#L230 `{vm.domarch}-3.0-unknown` is not a valid VirtualSystemType, so a file containing this would produce a warning on validation in https://github.com/xenserver/xenadmin/blob/bfc8141391bb892608d6a6d16150ad1401e68c2e/XenOvfApi/Validation.cs#L138. Have we seen such warning?

Reply:
No, because none of xenserver's templates or VMs seem to have a value for this field (with the exception of dom0, which is not exportable). Not sure if this is a toolstack bug or not. (By the way, the knownVirtualSystemTypes needs updating).